### PR TITLE
Added SCLI-21, SCLI-22, SCLI-37

### PR DESCRIPTION
**Pull request by metinavc1**
SCLI-21 Ability to suppress Merchant Password from POS for refund
SCLI-22 Allow surcharging from POS
SCLI-37 Automated Address Resolution - Implement device lookup for IP address using Serial Number

### CODE REVIEW

File: Library/SPIClient.h [hunk 1]
```diff
@@ -102,6 +107,29 @@ typedef void (^SPICompletionState)(BOOL alreadyMovedToIdleState, SPIState *state
  */
 @property (nonatomic, copy) NSString *posId;
 
+/**
+ Uppercase AlphaNumeric string that Indentifies your POS instance. This value
+ is displayed on the EFTPOS screen. Can only be called set in the Unpaired
+ state.
+ */
+@property (nonatomic, retain) NSString *serialNumber;
```
Comment (chrishulbert) [inline Library/SPIClient.h]: If you're feeling paranoid, you should change retain to copy here, to protect against api consumers setting a NSMutableString but no big deal.

File: Library/Service/SPIDeviceService.m [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+{
+    self = [super init];
+    
+    if (self) {
+        NSError *error = nil;
+        NSData *JSONData = [JSONString dataUsingEncoding:NSUTF8StringEncoding];
+        NSDictionary *JSONDictionary = [NSJSONSerialization JSONObjectWithData:JSONData options:0 error:&error];
+        
+        if (!error && JSONDictionary) {
+            for (NSString* key in JSONDictionary) {
```
Comment (chrishulbert) [inline Library/Service/SPIDeviceService.m]: Not sure why a for loop is needed here, typically you can just do `_address = JSONDictionary[@"ip"]` and the same for lastupdated.

File: Library/Service/SPIDeviceService.m [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
+    [request setHTTPMethod:@"GET"];
+    [request setValue:apiKey forHTTPHeaderField:@"ASM-MSP-DEVICE-ADDRESS-API-KEY"];
+    [request setURL:[NSURL URLWithString:deviceAddressUrl]];
+    
+    NSError *error = nil;
+    NSHTTPURLResponse *responseCode = nil;
+    
+    NSData *oResponseData = [NSURLConnection sendSynchronousRequest:request returningResponse:&responseCode error:&error];
```
Comment (chrishulbert) [inline Library/Service/SPIDeviceService.m]: Should probably make it clear somewhere that this function does synchronous IO and may take a while to return, js-style async io with callbacks is more typical in iOS, callers might be surprised that this takes a while to return

File: Tests/Model/SPITransactionTests.m [hunk 4]
```diff
@@ -141,6 +166,15 @@ - (void)testInitiateMotoPurchaseRequest {
     }];
 }
 
+- (void)testInitiateMotoPurchaseRequestWithSurchargeAmount {
+    SPIClient *client = [SPITestUtils clientWithTestSecrets];
+    
+    client.state.status = SPIStatusPairedConnected;
+    [client initiateMotoPurchaseTx:@"test" amountCents:19 surchargeAmount:1 completion:^(SPIInitiateTxResult *result) {
+        XCTAssertNotNil(result);
```
Comment (chrishulbert) [inline Tests/Model/SPITransactionTests.m:174]: Async unit tests in ios need to use XCTestExpectation or you can get all sorts of weirdness, like a subsequent test failing when this test was the culprit, see here for how to do it: https://developer.apple.com/documentation/xctest/asynchronous_tests_and_expectations/testing_asynchronous_operations_with_expectations

File: Tests/Model/SPITransactionTests.m [hunk 5]
```diff
@@ -150,6 +184,15 @@ - (void)testInitiateCashOutOnlyRequest {
     }];
 }
 
+- (void)testInitiateCashOutOnlyRequestWithSurcharge {
+    SPIClient *client = [SPITestUtils clientWithTestSecrets];
+    
+    client.state.status = SPIStatusPairedConnected;
+    [client initiateCashoutOnlyTx:@"test" amountCents:10 surchargeAmount:1 completion:^(SPIInitiateTxResult *result) {
+        XCTAssertNotNil(result);
```
Comment (chrishulbert) [inline Tests/Model/SPITransactionTests.m:192]: Again, use XCTestExpectation: https://developer.apple.com/documentation/xctest/asynchronous_tests_and_expectations/testing_asynchronous_operations_with_expectations

**Review by chrishulbert — CHANGES_REQUESTED** [review]
Just needs those things tidied up and good to go :)

File: Library/SPIClient.h (hunk 1 above)
Comment (metinavc1) [reply Library/SPIClient.h]: it is fixed

File: Library/Service/SPIDeviceService.m (hunk 2 above)
Comment (metinavc1) [reply Library/Service/SPIDeviceService.m]: Fixed

File: Library/Service/SPIDeviceService.m (hunk 3 above)
Comment (metinavc1) [reply Library/Service/SPIDeviceService.m]: Done

File: Tests/Model/SPITransactionTests.m (hunk 4 above)
Comment (metinavc1) [reply Tests/Model/SPITransactionTests.m:174]: This test is about Transactions via websocket. Previous test methods were developed like that. Tech debt todo

File: Tests/Model/SPITransactionTests.m (hunk 5 above)
Comment (metinavc1) [reply Tests/Model/SPITransactionTests.m:192]: This test is about Transactions via websocket. Previous test methods were developed like that. Tech debt todo

**Review by chrishulbert — APPROVED** [review]
Looks good, there's a little tech debt re existing async tests that might be worth looking at in future, cheers 🍻

**Pull request merged**